[doc] Mandate C11 in the C/C++ Coding Style Guide

lenary opened the pull request:
This matches the C version available when C++14 was standardised.

It is prudent to avoid nonstandard language extensions so that using
different compilers is easier. The style guide now includes a short
whitelist of allowed nonstandard extensions, and recommendations around
ensuring the usage of these can be made portable in future.

Closes #981

lenary commented:
I plan to move any software to C11 in a separate pull request.

imphil commented on doc/rm/c_cpp_coding_style.md:
```diff
@@ -173,6 +186,49 @@ This may change if we find cases where this initialization improves readability.
+*   `section(<name>)` to put a definition into a given object section.
+*   `weak` to make a symbol definition have weak linkage.
+*   `naked` to ensure a function has no prolog or epilog.
+*   `packed` to ensure a struct contains no padding.
```
Please also add `interrupt` to this list.

mcy replied:
You probably also want "aligned" or whatever GCC calls its explicit alignment attribute.

Also, maybe include "alias"? It can be useful for making stuff like `memcpy` link without actually having to define `memcpy` (e.g. if we define `ot_memcpy` instead).

imphil commented on doc/rm/c_cpp_coding_style.md:
````diff
@@ -173,6 +186,49 @@ This may change if we find cases where this initialization improves readability.
+void _crt(void) OT_SECTION(".crt");
+```
+
+This ensures changes to add compatibilty for other compilers are less invasive.
````
What compilers other than GCC and clang do we expect? This macro seems to make code harder to understand for the vague possibility that we might want to support another compiler in the future.

mcy replied:
Very unlikely that we'll be using other compilers; that said, I support hiding `__attribute__`s behind macros when possible, because:
- Frankly, it's an ugly GNUism that makes code hard to read.
- It's good hygine.

Both of those are kind of subjective reasons, but ultimately I think style guide arbitration is pretty subjective... =P

gkelly replied:
> What compilers other than GCC and clang do we expect? This macro seems to make code harder to understand for the vague possibility that we might want to support another compiler in the future.

I agree, I think `OT_SECTION` clouds the meaning.  It's not OT-specific but invites the reader to make that connection unless they're already familiar with the macros you're proposing.  Given the constraints that the attributes must already be GCC and Clang compatible I don't think putting them behind macros is valuable.

lenary replied:
Yeah, I'm swayed that it's easier to google for `__attribute__((section(".crt")))` than `OT_SECTION`. Yes it's a little uglier, but it's also unambiguous. We know that `__attribute__(…)` is supported by both GCC and Clang. 

I'm happy to revisit the issue when we want to support even more compilers, which are not compatible with the existing attribute syntax. I don't think we're at that point yet.

mcy commented on doc/rm/c_cpp_coding_style.md:
````diff
@@ -173,6 +186,49 @@ This may change if we find cases where this initialization improves readability.
+```c
+#define OT_SECTION(name) __attribute__((section(name)))
+//...
+void _crt(void) OT_SECTION(".crt");
````
Should we specify where the attributes go? I would actually suggest
```c
OT_SECTION(".crt")
void _crt(void);
```
This has the advantages of:
- Working regardless of being applied to a function prototype or a function definition.
- It helps stop the line width from getting totally unreasonable.

Mind, we don't really need to put it on its own line, but at least putting it at the start is a good idea.

mcy commented on doc/rm/c_cpp_coding_style.md:
````diff
@@ -173,6 +186,49 @@ This may change if we find cases where this initialization improves readability.
+For instance, `__builtin_abort()` should not be called.
+Instead, the call should be hidden by a compatible definition, like below.
+```c
+_Noreturn void abort(void) {
````
We're going to have stdnoreturn available, so maybe just use `noreturn`?

mcy commented on doc/rm/c_cpp_coding_style.md:
```diff
@@ -173,6 +186,49 @@ This may change if we find cases where this initialization improves readability.
+This ensures changes to add compatibilty for other compilers are less invasive.
+
+To avoid overhead, this function can be declared as an `inline` function in a header file.
+If it is declared as an inline function, exactly one compilation unit must have a compatible `extern inline` declaration for that function.
```
Maybe include an example? C `inline` tends to be really finicky to get right.

lenary replied:
Will do.

lenary commented:
I have added some further guidance on compiler builtin usage (based on a discussion with @asb), as it's a tricky issue to get correct. The fundamental point is that someone from the compiler team should be looped in on reviewing any changes that use builtins.

imphil commented on doc/rm/c_cpp_coding_style.md:
````diff
@@ -173,6 +186,58 @@ This may change if we find cases where this initialization improves readability.
+Nonstandard attributes must be written before the declaration, like the following example, so they work with both declarations and definitions.
+
+```c
+__attribute__((section(".crt"))) void _crt(void);
````
Can you please file a follow-up to get existing uses in the code base fixed to match this style?

lenary replied:
Yep, there'll be quite a significant PR which upgrades existing C code to C11. I'll start on that tomorrow.

lenary commented:
I have removed the explicit allow rule for `__attribute__((naked))`, after a short discussion on #1142 and ensuring we don't have any in the codebase at the moment. This doesn't mean `naked` is banned, it means we now recommend against it.

I think I meant to write `interrupt` originally anyway, and the lack of `interrupt` was already noted by @imphil, and corrected.

asb commented on doc/rm/c_cpp_coding_style.md:
```diff
@@ -143,6 +143,19 @@ The [Google C++ Style Guide](https://google.github.io/styleguide/cppguide.html)
+
+The following nonstandard extensions may be used:
+*   Inline assembly
+*   Nonstandard Attributes
```
Nit: for this and the next line, Attributes and Builtins surely don't need to be capitalised? Or alternatively, capitalise "assembly" in the line above for consistency.

lenary replied:
Updated
